src/chunk.c

```diff
@@ -3755,7 +3755,15 @@ ts_chunk_drop_internal(const Chunk *chunk, DropBehavior behavior, int32 log_leve
 void
 ts_chunk_drop(const Chunk *chunk, DropBehavior behavior, int32 log_level)
 {
-	ts_chunk_drop_internal(chunk, behavior, log_level, false);
+	PG_TRY();
+	{
+		ts_chunk_drop_internal(chunk, behavior, log_level, false);
+	}
+	PG_CATCH();
+	{
+		return;
+	}
+	PG_END_TRY();
```

Comment:
This is used by a lot of other functions that are not part of the policy execution, meaning that you ignore the error when, for example, a user calls `drop_chunk` directly. You cannot capture the error at this level, you need to capture this in the policy execution code.

Reply:
Please check my analysis:

```
1. ts_chunk_drop_internal (/Volumes/Work/timescaledb/src/chunk.c:3737)
2. ts_chunk_do_drop_chunks (Unknown Source:0)
3. ts_chunk_drop_chunks (/Volumes/Work/timescaledb/src/chunk.c:4212)
4. ExecMakeFunctionResultSet (/Volumes/Work/postgresql-14.0/src/backend/executor/execSRF.c:624)
5. chunk_invoke_drop_chunks (/Volumes/Work/timescaledb/tsl/src/chunk.c:261)
6. policy_retention_execute (/Volumes/Work/timescaledb/tsl/src/bgw_policy/job.c:282)
```

Above call stack for policy retention code flow:
policy_retention_execute() is the function which starts executing the retention policy. As part of this policy we find all eligible chunks which needs to be dropped. Once number of chunks are obtained, we loop through these chunks and drop one by one.

Check frame 3 above in code.
Call to ts_chunk_drop_chunks() is embedded in PG_TRY() and PG_CATCH() block.
In frame 2 where ts_chunk_do_drop_chunks() is executed, this function will check for all eligible chunks and looping through each chunk, and calling ts_chunk_drop_internal() on the eligible chunk.

Now any error reported in further calls after ts_chunk_drop_internal() will unwind the stack till frame 3 and policy is exited with an error.

frame 2 is the place where we loop through each chunk. Since there was an exception (aka error) this frame is unwinded and we loose all context to proceed to next chunks.
Thus the fix was to embed frame 1 in PG_TRY() and PG_CATCH() block. So that we can go back to the loop which fetches for next chunk and proceed with it.

Reply:
As discussed I will have a new issue to fix for data retention policy.